src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentStates.cs

```diff
@@ -270,4 +295,44 @@ public async ValueTask<ChecksumsAndIds<DocumentId>> GetChecksumsAndIdsAsync(Canc
         var documentChecksums = new ChecksumCollection(await documentChecksumTasks.WhenAll().ConfigureAwait(false));
         return new(documentChecksums, SelectAsArray(static s => s.Id));
     }
+
+    public void AddDocumentIdsWithFilePath(ref TemporaryArray<DocumentId> temporaryArray, string filePath)
+    {
+        // Lazily initialize the file path map if not computed.
+        _filePathToDocumentIds ??= ComputeFilePathToDocumentIds();
+
+        if (_filePathToDocumentIds.TryGetValue(filePath, out var oneOrMany))
+        {
+            foreach (var value in oneOrMany)
+                temporaryArray.Add(value);
+        }
+    }
+
+    public DocumentId? GetFirstDocumentIdWithFilePath(string filePath)
+    {
+        // Lazily initialize the file path map if not computed.
+        _filePathToDocumentIds ??= ComputeFilePathToDocumentIds();
+
+        return _filePathToDocumentIds.TryGetValue(filePath, out var oneOrMany)
+            ? oneOrMany.FirstOrDefault()
+            : null;
+    }
+
+    private FrozenDictionary<string, OneOrMany<DocumentId>> ComputeFilePathToDocumentIds()
+    {
+        using var _ = PooledDictionary<string, OneOrMany<DocumentId>>.GetInstance(out var result);
```

Comment:
>OneOrMany [](http://example.com/codeflow?start=47&length=9)

I'd like to better understand the decision to use OneOrMany as both the storage type and intermediary type in the calculation. Let's chat in person.

Reply:
so i couldn't find any invariants stating that within a single TextDocumentStates filepaths are unique (though i really hope they are).  So i wrote the code to be resilient to the filepaths not being unique and duplicates being possible.

However, that should be exceedingly rare.  As such, i use a OneOrMany so that we don't allocate ImmutableArrays for literally every filepath in the world.  Instead, we just have a struct, that normally poitns directly at the single DocumentId for that filepath.